runtime/src/main/java/org/corfudb/runtime/collections/StreamNotificationTask.java

```diff
@@ -84,7 +92,13 @@ private void sendNotifications() throws Exception {
 
             // Send notification to client with the pre-registered callback.
             startTime = endTime;
-            subscription.getStreamingMetrics().recordDeliveryDuration(() -> listener.onNext(nextUpdate));
+            try {
+                subscription.getStreamingMetrics().recordDeliveryDuration(() -> listener.onNext(nextUpdate));
+            } catch (Throwable t) {
```

Comment:
The only Throwable expected from Corfu's POV is a TrimmedException. If Throwable is not an instance of TrimmedException then it's user exception.

Reply:
I was not 100% sure this is true. I agree in the theory it should, but I felt I might be missing on something. For instance, I just experimented with the test and if the runtime is shutdown after the subscription is done we'll get on the onError the following:

org.corfudb.runtime.exceptions.UnrecoverableStreamingException: java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.CompletableFuture$AsyncSupply@5326d2e3 rejected from java.util.concurrent.ThreadPoolExecutor@2c0da708[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 1]

Which is a Corfu exception, of course the issue here is major, as the runtime is disconnected but then it kind of breaks the assumption and like this I'm not sure if we could be missing something else.

Reply:
Those are generic Java exceptions and I think its fine to pass them to the client. Not only can `RejectedExecutionException`, but also `InterruptedException` both of which are hard failures, wrapping them doesn't add any value, or facilitate for processing downstream. 

The point of `StreamSubscriptionException` is to signal to the client a `soft error` that is retriable, but this patch clubs two different classes of errors (soft/hard) in a single exception. 

Furthermore, those exceptions happen on daemon threads that we control, so if they happen they will not prevent other client logic from processing. There is a use case where clients would want to do post processing after shutting down the StreamingManager and those APIs are already available.

Reply:
StreamSubscriptionException essentially should only wrap `TrimmedException`, that can be thrown directly too. But later on down the road there might be multiple soft errors that are retriable and StreamSubscriptionException can be an indirection to wrap those exceptions as well.

Reply:
Just to be sure we're on the same page (as you wrote it in the opposite way) exceptions wrapped as 'StreamSubscriptionException' are actually <NOT> retriable. The message to consumers will be, if you receive a 'StreamSubscriptionException' do not retry, get a snapshot and then re-attach. Any other type of exception you can attempt to re-attach at your own risk.

Reply:
@Maithem addressed all comments, please re-review and let me know if you see it this way too (previous comment)... I've already changed it such that TrimmedException is the only Corfu expected exception in the NotificationTask, though I don't think we should ever hit this, cause TrimmedException could be hit only on the pollingTask but not on the NotificationTask (as by the time of notification the data has already been read and its enqueued)

Reply:
I actually removed it, the entry has already been read and is enqueued... there is no way we can hit a TrimmedException...